Add access to the Peers tab from the network icon

hebasto opened the pull request:
This PR add a small context menu to the network activity icon that provides an access to the Peers tab:

![gui-network-icon](https://user-images.githubusercontent.com/32963518/116794314-d64b9b80-aad4-11eb-89ca-7f75c7442ba8.gif)

Closes #93.

kristapsk commented on src/qt/bitcoingui.cpp:
```diff
@@ -933,9 +938,24 @@ void BitcoinGUI::setNumConnections(int count)
+            //: A context menu item.
+            tr("Disable network activity") :
+            //: A context menu item. The network activity was disabled previously.
+            tr("Enable network activity again"),
```
Maybe not so important, but I think "again" is redundant here, could be just "Enable network activity".

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin-core/gui/pull/309#issuecomment-830763475).

hebasto commented:
Updated 1f208f4360df55f18afb89056eb4c18e10f9fd5f -> a7f5ea589fb7a6eddca3e060eade403b2dd6ff87 ([pr309.02](https://github.com/hebasto/gui/commits/pr309.02) -> [pr309.03](https://github.com/hebasto/gui/commits/pr309.03), [diff](https://github.com/hebasto/gui/compare/pr309.02..pr309.03)):

- addressed @kristapsk's [comment](https://github.com/bitcoin-core/gui/pull/309#discussion_r624580987)

kristapsk reviewed (approved):
ACK a7f5ea589fb7a6eddca3e060eade403b2dd6ff87

jarolrod reviewed (commented):
tACK a7f5ea589fb7a6eddca3e060eade403b2dd6ff87

Tested on macOS 11.3 Qt 5.15.2, Also tested with `-disablewallet=1`.

This is ok to merge as-is. But, wanted to document how I think this should be changed around as part of a follow-up.

The 'Peers Tab' is for seeing your connected Peers and interacting with them, such as banning or disconnecting a peer.

The idea of "Disabling Network Activity" should be tied to the 'Network Traffic Tab' not the 'Peers Tab'. This means we could add a "Disable Network Activity" button to the Network Traffic tab and a 'Network Traffic' widget next to the 'Connection Count' widget. Pressing this 'Network Traffic' widget would open the 'Network Traffic' tab where the user has the option of disabling network activity.

hebasto commented:
@Sjors Do you mind looking into this PR?

Sjors reviewed (approved):
tACK a7f5ea589fb7a6eddca3e060eade403b2dd6ff87

promag commented on src/qt/bitcoingui.cpp:
```diff
@@ -586,7 +584,10 @@ void BitcoinGUI::setClientModel(ClientModel *_clientModel, interfaces::BlockAndH
-        updateNetworkState();
+        setNetworkActive(m_node.getNetworkActive());
+        connect(connectionsControl, &GUIUtil::ClickableLabel::clicked, [this] {
+            m_network_context_menu->exec(QCursor::pos());
```
`QMenu::exec` is blocking. If you want to keep this approach then you could declare a local `QMenu` instance instead - this way you can ditch the class member and the call to `.clear()`.

hebasto replied:
@promag 

Thanks!

> `QMenu::exec` is blocking.

My bad! [Fixed](https://github.com/bitcoin-core/gui/pull/309#issuecomment-835154862).

> If you want to keep this approach then you could declare a local `QMenu` instance instead - this way you can ditch the class member and the call to `.clear()`.

How the lifetime of local `QMenu` instances could be managed in a proper way?

promag reviewed (commented):
Tested ACK a7f5ea589fb7a6eddca3e060eade403b2dd6ff87.

hebasto commented:
Updated a7f5ea589fb7a6eddca3e060eade403b2dd6ff87 -> d29ea72393ac1d9b32a6976062e9c9fb75876295 ([pr309.03](https://github.com/hebasto/gui/commits/pr309.03) -> [pr309.04](https://github.com/hebasto/gui/commits/pr309.04), [diff](https://github.com/hebasto/gui/compare/pr309.03..pr309.04)):

- addressed @promag's [comment](https://github.com/bitcoin-core/gui/pull/309#discussion_r626571342):
> `QMenu::exec` is blocking.

RandyMcMillan commented:
It would be great if a key combo was added to this functionality...
Maybe:

COMMAND + OPTION <w> or <n> would work.

or

Maybe COMMAND + OPTION <o> would work.

kristapsk commented:
> Adding a shortcut for enabling/disabling the network activity

I'm not even sure such shortcut is a good idea at all. Could accidentally hit that, the same that has happened to more with that icon without this PR.

promag reviewed (commented):
Code review ACK d29ea72393ac1d9b32a6976062e9c9fb75876295.

Not fond of the approach though. It seems it would be enough to update the menu item text on `BitcoinGUI::updateNetworkState()` instead of `QMenu::clear()` + `QMenu::addAction()` and the slot would just query the current network state.

It also seems more elegant to just call `BitcoinGUI::updateNetworkState()` in `BitcoinGUI::setClientModel` which already calls `m_node.getNetworkActive()`.

Worth noting that when the toggle action is triggered it calls `m_node.setNetworkActive()` which results in destroying that same action due to `QMenu::clear()`.

Don't mind my feedback, it's just another way of implementing the feature, thanks for adding the feature and addressing my previous comment.

Sjors commented:
re-ACK d29ea72393ac1d9b32a6976062e9c9fb75876295
